File: Jasmine/JasmineCompiler/Src/DFIR_Compiler/DFIR_Compiler/IFeedbackOutputNode.cs
using System;
using System.Collections.Generic;


namespace DFIR_Compiler
{
    public class IFeedbackOutputNode:INode
    {
        private InputNode inputNode;

        public IFeedbackOutputNode(int nodeId, int parentId, List<ITerminal> inputTerminals, List<ITerminal> outputTerminals, InputNode _inputNode) : base(nodeId, parentId, inputTerminals, outputTerminals)
        {
            inputNode = _inputNode;
            INodeType = "IFeedbackOutputNode";
            indegree++; //Because the IFeedbackOutputNode has the vitual connection with IFeedbackInputNode
        }

        public virtual InputNode GetInputNode()
        {
            return inputNode;
        }
        public virtual IFeedbackInputNode GetIFeedbackInputNode()
        {
            return (IFeedbackInputNode)GetIdToINode()[inputNode.GetNodeId()];
        }
        public virtual string GetInitName()
        {
            if (GetIsInputConnected())
            {
                return "Init" + Convert.ToString(nodeId);
            }
            return "0";
        }

        public virtual void SetInputNode(InputNode _inputNode)
        {
            inputNode = _inputNode;
        }


        public virtual string GetLoopIndexName()
        {
            if (GetParentINode().GetParentINode() != null)
            {
                if (GetParentINode().GetParentINode().GetINodeType() == "IForLoop")
                {
                    IForLoop f = (IForLoop)GetParentINode().GetParentINode();
                    return f.GetILoopIndex().GetName();
                }
                else
                {
                    return "";
                }
            }
            else
            {
                return "";
            }

        }

        public override void Print()
        {
            Console.WriteLine("------------------------------------");
            Console.WriteLine("IFeedbackOutputNode:");
            Console.WriteLine("NodeId: " + nodeId + ' ' + "ParentId: " + parentId);
            inputNode.Print();
            for (int i = 0; i < inputTerminals.Count; i++)
            {
                inputTerminals[i].Print();
            }
            for (int i = 0; i < outputTerminals.Count; i++)
            {
                outputTerminals[i].Print();
            }
        }

    }
}

File: Wayne/WayneCompiler/src/Dfir_representation/IFeedbackOutputNode.java
package Dfir_representation;

import java.util.Vector;


public class IFeedbackOutputNode extends JNode{
	   
    
       private InputNode inputNode;
	   public IFeedbackOutputNode(int nodeId, int parentId, Vector<ITerminal> inputTerminals,Vector<ITerminal> outputTerminals,InputNode inputNode){
		   super(nodeId,parentId,inputTerminals,outputTerminals);
		   this.inputNode = inputNode;
		   JNodeType = "IFeedbackOutputNode";
	   }
	  
	   public void print(){
		   System.out.println("IFeedbackOutputNode begin");
		   System.out.println("NodeId: " + this.nodeId + ' ' + "ParentId: " + this.parentId);
		   this.inputNode.print();
		   for(int i=0; i<inputTerminals.size(); i++){
				   inputTerminals.get(i).print();
		   }
		   for(int i=0; i<outputTerminals.size(); i++){
				   outputTerminals.get(i).print();
		   }   
		  System.out.println("IFeedbackOutputNode end\n");
	   }
	 

    public IFeedbackInputNode getIFeedbackInputNode(){
    	return (IFeedbackInputNode)JNode.getIdToJNode().get(inputNode.getNodeId());
    }
	public InputNode getInputNode() {
		return inputNode;
	}

	public void setInputNode(InputNode inputNode) {
		this.inputNode = inputNode;
	}
	
	public String getInitName(){
		if(this.getIsInputConnected()){
			return "Init"+Integer.toString(this.nodeId);
		}
		return "0";
	}
	public String getLoopIndexLabel(){
		if(this.getParentJNode().getParentJNode() != null){
			if(this.getParentJNode().getParentJNode().getJNodeType() == "IForLoop"){
				IForLoop f = (IForLoop) this.getParentJNode().getParentJNode();
				return f.getILoopIndex().getName();
			}else{
				return "";
			}
		}else{
			return "";
		}
		
	}
}
